zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -530,6 +531,17 @@ public void takeSnapshot(boolean syncSnap) {
         ServerMetrics.getMetrics().SNAPSHOT_TIME.add(elapsed);
     }
 
+    public boolean existSnapshot() {
+        File snapShotDir = txnLogFactory.getSnapDir();
+        String[] snaps = snapShotDir.list(new FilenameFilter() {
+            @Override
+            public boolean accept(File dir, String name) {
+                return name.contains("snapshot");
```

Comment:
- `Util.isSnapshotFileName(name)` is better?
- Just find one valid snapshot, then fast return true instead of iterating all?
- I still need a deeper thinking about this change, will come back soon. Sorry for the late.

Reply:
- Sometimes the `DIFF` message is frequent,  so the existing check of snapshot may be not cheap? Putting that check just before the real place of `takeSnapshot` is more efficiency?
- Another alternative way may be: allow users to take a snapshot themselves when upgrading by CLI or admin server command(take a snapshot immediately) that were [PR-917](https://github.com/apache/zookeeper/pull/917) and [PR-1044 ](https://github.com/apache/zookeeper/pull/1044) wanted to do.

Reply:
Furthermore, we should answer this question: Why must we need a snapshot to restore at the start-up of server(a bit of a counter-intuitive)? we can remove this restriction without any consistency concerns?

Reply:
1. this check only occurr during qurumpeer starting.
2.  the reason of check snapshot whether exist is prevent zk from starting in a subset and maybe become a leader.